src/uarch.rs

```diff
@@ -0,0 +1,380 @@
+    Puma,
+    Bulldozer,
+    Piledriver,
+    SteamRoller,
+    Excavator,
+    // Zen,
+    Zen,
+    ZenPlus, // PinnacleRidge, Picasso,
+    Zen2,
+    Zen3,
+    Zen4,
+}
+
+#[non_exhaustive]
+pub enum Core {
+    Homogenous(CoreArch),
+    Heterogeneous { P: CoreArch, E: CoreArch },
+    // Note : This doesn't work for more complex set-up with three type of cores.
+    // Such design do not exist in Intel (ISA) land, but exist in ARM land.
+}
+
+#[non_exhaustive]
+pub enum UArch {
+    // ---- Intel ----
+    // i486
+    i486, // Note, this may eventually be split,
+    // Family 5
+    // P5
+    P5, // 80501 and P54C aka 80502
+    P5MMX,  // 80503 (Pentium MMX)
+
+    // NetBurst aka P68
+    Willamette,
+    Northwood,
+    Prescott,
+    CedarMill, // Prescott Die Shrink
+    // P6
+    P6PentiumPro,
+    P6PentiumII,
+    P6PentiumIII,
+    // Pentium M
+    Banias,
+    Dothan,
+    Tolapai,
+    // Modified Pentium M
+    Yonah,
+    Merom,
+    Penryn,
+    Nehalem,
+    Westmere,
+    SandyBridge,
+    IvyBridge,
+    IvyBridgeE,
+    Haswell,
+    HaswellE,
+    Broadwell,
+    Skylake,
+    SkylakeServer,
+    KabyLake,
+    CascadeLake,
+    CoffeeLake,
+    CooperLake,
+    CannonLake,
+    WhiskeyLake,
+    AmberLake,
+    CometLake,
+    IceLake,
+    IceLakeServer,
+    TigerLake,
+    RocketLake,
+    AlderLake,
+    SapphireRapids,
+    RaptorLake,
+    EmeraldRapids,
+    MeteorLake,
+
+    // Small Cores
+    // No split for now.
+    Bonnel,
+    Saltwell,
+    Silvermont,
+    Airmont,
+    Goldmont,
+    GoldmontPlus,
+    Tremont,
+    Gracemont,
+    Crestmont,
+
+    // Lakemont
+    Quark,
+    // Many Core / Xeon Phi (TODO)
+    // Polaris is not x86 ISA
+    // Larrabee
+    // RockCreek,
+    KnightsFerry,
+    KnightsCorner,
+    KnightsLanding,
+    KnightsMill,
+
+    // ---- AMD ----
+    Am486,
+    EnhancedAm486,
+    Am586,
+    K5,
+    K6,
+    K6_2,
+    K6III,
+    K7,
+    K8, // Hammer
+    // K9 was never finished.
+    K10, // Llano
+    Bobcat,
+    Jaguar,
+    // Enhanced Jaguar for XboX One X 'Scorpio". (CPUID currently unknown)
+    Puma,
+    Bulldozer,
+    Piledriver,
+    SteamRoller,
+    Excavator,
+    // Zen,
+    Zen,
+    HygonDhyana,
+    ZenPlus, // PinnacleRidge, Picasso,
+    Zen2,
+    Zen3,
+    Zen4,
+
+}
+
+pub struct MicroArchitecture {
+    pub vendor: Vendor,
+    pub cores: Core,
+    pub codename: UArch,
+    // This currently leaves off the table the exact variant (e.g Amber Lake U).
+}
+
+
+// Source for the tables :
+//
+// Intel Software Developper Manual (and change document)
+//
+// https://en.wikichip.org/wiki/intel/cpuid
+// https://en.wikichip.org/wiki/amd/cpuid
+//
+// https://en.wikipedia.org/wiki/List_of_Intel_CPU_microarchitectures
+// https://en.wikipedia.org/wiki/List_of_AMD_CPU_microarchitectures
+
+
+// ===================
+// Micro-Architectures
+// ===================
+const INTEL_486: MicroArchitecture = MicroArchitecture {
```

Comment:
maybe all this could be parsed at build time from some csv file and generated with the  `phf` library or similar?

Reply:
(not strictly necessary of course, just wondering if it maybe makes things easier)

Reply:
I definitely remarked that my design required a lot of boilerplate that could be autogenerated.